Sources/Subs-Admin.php

```diff
@@ -759,197 +794,201 @@ function updateSettingsFile($config_vars, $keep_quotes = null, $partial = false)
+		if (!empty($setting_def['auto_delete']) && empty($config_vars[$var]) && empty($GLOBALS[$var]) && in_array($var, array_keys($GLOBALS)))
+			$config_vars[$var] = $setting_def['default'];
 
-		$flags = !empty($context['utf8']) || !empty($utf8) ? 'u' : '';
-		switch (gettype($var))
+		if (!isset($config_vars[$var]))
 		{
-			case 'string':
-				$regex = '
-						# If the string starts with a single or double quotation mark
-						(?(?=["\'])
-							# match the opening quotation mark...
-							(["\'])
-							# then any number of other characters or escaped quotation marks...
-							(?:[^\\1]|(?<=\\\)\\1)*
-							# then the closing quotation mark.
-							\\1
-						)';
-				$flags .= 'x';
-				break;
-
-			case 'integer':
-				$regex = '[+-]?\d+';
-				break;
-
-			case 'double':
-				$regex = '[+-]?\d+\.\d+';
-				break;
-
-			case 'boolean':
-				$regex = '(?i:TRUE|FALSE)';
-				break;
-
-			case 'NULL':
-				$regex = '(?i:NULL)';
-				break;
-
-			case 'array':
-				// This uses a PCRE subroutine to match nested arrays.
-				$regex = 'array\s*(\((?>[^()]|(?1))*\))';
-				break;
-
-			default:
-				$regex = '';
-				break;
+			// If a defined setting is missing from the file, we need to fix that ASAP.
+			if (!in_array($var, array_keys($GLOBALS)) && empty($setting_def['required']))
+				$config_vars[$var] = $setting_def['default'];
+
+			// If we're rebuilding, we want (almost) all of them.
+			elseif (!empty($rebuild) && (!isset($setting_def['auto_delete']) || !empty($GLOBALS[$var])))
+				$config_vars[$var] = $GLOBALS[$var];
+
+			// Otherwise, just leave this one alone.
+			else
+				unset($settings_defs[$var]);
 		}
+	}
 
-		return array($regex, $flags);
-	};
+	/******************************
+	 * PART 2: Content processing *
+	 ******************************/
 
-	// Use var_export to make sure the value is formatted correctly,
-	// but customize the output to match SMF coding conventions.
-	$pretty_var_export = function($var)
-	{
-		if (is_array($var))
-			return preg_replace_callback('/^\h+/m', function ($m) { return strtr($m[0], array('  ' => "\t")); }, var_export($var, true));
-		else
-			return var_export($var, true);
-	};
+	$type_regex = array(
+		'string' =>
+			# If the string starts with a single or double quotation mark
+			'(?(?=["\'])' .
+				# match the opening quotation mark...
+				'(["\'])' .
+				# then any number of other characters or escaped quotation marks...
+				'(?:[^\\1]|(?<=\\\)\\1)*?' .
+				# then the closing quotation mark.
+				'\\1' .
+			')',
+		// Some integer values might have been stored as strings.
+		'integer' =>  '["\']?[+-]?\d+["\']?',
+		'double' =>  '[+-]?\d+\.\d+([Ee][+-]\d+)?',
+		'boolean' =>  '(?i:TRUE|FALSE)',
+		'NULL' =>  '(?i:NULL)',
+		// These use a PCRE subroutine to match nested arrays.
+		'array' =>  'array\s*(\((?>[^()]|(?1))*\))',
+		'object' =>  '\w+::__set_state\(array\s*(\((?>[^()]|(?1))*\))\)',
+	);
 
-	// Time to build our new Settings.php!
+	/**
```

Comment:
one asterisk here please so it can be a regular multi-line comment and not a docblock.